### PR TITLE
Audit Fixes EIP2930/1559 RLP Decoding Table Correspondence

### PR DESCRIPTION
**Pull request by darth-cy**
### Description

This PR resolves a major issue in RLP audit that raises concern for malicious RLP stack op injection into the decoding table. The fix includes lookups from the sorted decoding table into `rlp_table` and RLP state machine to ensure row correspondence.

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: zkevm-circuits/src/rlp_circuit_fsm.rs [hunk 1]
```diff
@@ -2262,6 +2220,188 @@ impl<F: Field> RlpCircuitConfig<F> {
...
+        ///////////////////////////////////////////////////////////////////
+
+        // RLP Decoding Table is sorted using an id = (tx_id, format, depth, access_list_idx,
+        // storage_key_idx) but bytes in the RLP circuit (rlp_table) is processed in order.
+        // Therefore, to prevent malicious injection of stack ops that don't correspond to
+        // actual bytes in the RLP circuit, lookups are added to ensure correct correspondence.
+
+        // The Init Op is not affected by sorting and stays at first position in decoding table.
+        // For this reason, its correctness and correspondence is constrained using a gate instead
+        // of lookup.
```
Comment (kunxian-xia) [inline zkevm-circuits/src/rlp_circuit_fsm.rs]: why is there no lookup from RLP circuit into RLP decoding table(storing `remaining_bytes`)?

File: zkevm-circuits/src/rlp_circuit_fsm.rs [hunk 2]
```diff
@@ -2262,6 +2220,196 @@ impl<F: Field> RlpCircuitConfig<F> {
...
+                meta.query_advice(rlp_decoding_table.tx_id, Rotation::cur()),
+                meta.query_advice(rlp_decoding_table.format, Rotation::cur()),
+                meta.query_advice(rlp_decoding_table.byte_idx, Rotation::cur()),
+                1.expr(), // PUSH op increases depth
+                // List scenario on each depth.
+                // Sufficiency is achieved by listing all possible depth.
+                meta.query_advice(is_stack_depth_one, Rotation::cur()),
+                meta.query_advice(is_stack_depth_two, Rotation::cur()),
+                meta.query_advice(is_stack_depth_three, Rotation::cur()),
+                meta.query_advice(is_stack_depth_four, Rotation::cur()),
```
Comment (kunxian-xia) [inline zkevm-circuits/src/rlp_circuit_fsm.rs]: We need to make sure the number of bytes pushed into higher depth must match the one sm extracts from `BeginObject / BeginVector` header. For a tag = `BeginObject / BeginVector` that starts with byte in the interval `[0xc0, 0xf8)`, the table expr for value should be `byte_value - 0xc0`; for tag starts with byte that are larger than `0xf8`, the table expr for value should be `tag_value_acc`. Therefore it's more reasonable to do lookup from sm into RlpDecodingTable.

File: zkevm-circuits/src/rlp_circuit_fsm.rs [hunk 3]
```diff
@@ -2262,6 +2220,196 @@ impl<F: Field> RlpCircuitConfig<F> {
...
+        meta.lookup_any("Decoding table stack op PUSH correspondence", |meta| {
+            let enable = meta.query_advice(rlp_decoding_table.is_stack_push, Rotation::cur());
+
+            let input_exprs = vec![
+                meta.query_advice(rlp_decoding_table.tx_id, Rotation::cur()),
+                meta.query_advice(rlp_decoding_table.format, Rotation::cur()),
+                meta.query_advice(rlp_decoding_table.byte_idx, Rotation::cur()),
+                1.expr(), // PUSH op increases depth
+                // List scenario on each depth.
+                // Sufficiency is achieved by listing all possible depth.
```
Comment (kunxian-xia) [inline zkevm-circuits/src/rlp_circuit_fsm.rs]: rlp_decoding_table.value is missing.

File: zkevm-circuits/src/rlp_circuit_fsm.rs (hunk 2 above)
Comment (darth-cy) [reply zkevm-circuits/src/rlp_circuit_fsm.rs]: This correspondence is added in the latest commits.

File: zkevm-circuits/src/rlp_circuit_fsm.rs [hunk 4]
```diff
@@ -2256,6 +2226,227 @@ impl<F: Field> RlpCircuitConfig<F> {
...
+        meta.lookup_any("Decoding table stack op PUSH correspondence", |meta| {
+            let enable = meta.query_advice(is_push_op_lookup, Rotation::cur());
+
+            let input_exprs = vec![
+                meta.query_advice(tx_id, Rotation::cur()),
+                meta.query_advice(format, Rotation::cur()),
+                meta.query_advice(byte_idx, Rotation::cur()),
+                meta.query_advice(depth, Rotation::cur()) + 1.expr(),
+                1.expr(),
+                select::expr(
```
Comment (kunxian-xia) [inline zkevm-circuits/src/rlp_circuit_fsm.rs:2370]: Should you add `al_idx` in the input_exprs?

File: zkevm-circuits/src/rlp_circuit_fsm.rs [hunk 5]
```diff
@@ -2256,6 +2226,227 @@ impl<F: Field> RlpCircuitConfig<F> {
...
+        });
+
+        meta.lookup_any("Decoding table stack op UPDATE correspondence", |meta| {
+            let enable = meta.query_advice(is_update_op_lookup, Rotation::cur());
+
+            let input_exprs = vec![
+                meta.query_advice(tx_id, Rotation::cur()),
+                meta.query_advice(format, Rotation::cur()),
+                meta.query_advice(byte_idx, Rotation::cur()),
+                meta.query_advice(depth, Rotation::cur()),
```
Comment (kunxian-xia) [inline zkevm-circuits/src/rlp_circuit_fsm.rs:2437]: Should you add `al_idx` in the input exprs?

File: zkevm-circuits/src/rlp_circuit_fsm.rs (hunk 4 above)
Comment (darth-cy) [reply zkevm-circuits/src/rlp_circuit_fsm.rs:2370]: `al_idx` is now part of input_exprs.

File: zkevm-circuits/src/rlp_circuit_fsm.rs (hunk 5 above)
Comment (darth-cy) [reply zkevm-circuits/src/rlp_circuit_fsm.rs:2437]: `al_idx` is now part of input_exprs.

**Review by kunxian-xia — APPROVED** [review]

**Pull request merged**